src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticRangeVisitor.cs

```diff
@@ -40,11 +40,13 @@ public static List<SemanticRange> VisitAllNodes(RazorCodeDocument razorCodeDocum
             rangeAsTextSpan = range.AsRazorTextSpan(sourceText);
         }
 
-        var visitor = new TagHelperSemanticRangeVisitor(razorCodeDocument, rangeAsTextSpan, razorSemanticTokensLegend, colorCodeBackground);
+        using var _ = ArrayBuilderPool<SemanticRange>.GetPooledObject(out var builder);
```

Comment:
The resizing of this array shows up in the profile as being a slight problem, but with only one test run being profiled there is no pooling benefit, so hopefully in the real world it goes away.